pokernetwork/pokerserver.8
.\"
.\" Copyright (C) 2005, 2006 Mekensleep <licensing@mekensleep.com>
.\"                          24 rue vieille du temple, 75004 Paris
.\"       
.\" This software's license gives you freedom; you can copy, convey,
.\" propagate, redistribute and/or modify this program under the terms of
.\" the GNU Affero General Public License (AGPL) as published by the Free
.\" Software Foundation (FSF), either version 3 of the License, or (at your
.\" option) any later version of the AGPL published by the FSF.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU Affero General Public License
.\" along with this program in a file in the toplevel directory called
.\" "AGPLv3".  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Authors:
.\"  Loic Dachary <loic@dachary.org>
.\"
.TH pokerserver 8 local
.SH NAME
pokerserver \- multiplayer poker server

.SH SYNOPSIS
pokerserver

.SH DESCRIPTION

Run a multiplayer poker server

.SH AUTHORS

Loic Dachary <loic@gnu.org>, Cedric Pinson <cpinson@freesheep.org>,
Henry Precheur <henry@precheur.org>


NEWS
Release 2.0.0

Protocol is backward incompatible.

Remove poker2d and the pokerui library: they are now in a separate package.

Release 1.7.5

New feature : Satellite tournaments.
New feature : PacketPokerCreateTourney allows for user created tournaments.
New feature : TablePicker packet chose a table where a player can sit and play.
New feature : PacketPokerPlayerHandStrength returns the strengh of the hand
New feature : the prize currency for a tournament can be different from the
              buy_in currency
New feature : bots refill by default.
Backward incompatible change : /REST http endpoint no longer exists.
Server code has 100% coverage with tests (not including bots and client libraries).
Improve documentation of poker.server.xml.
Fix bug where pokerconfigupgrade fails to run.
Add [serial] to PokerExplain log messages for clarity.
Fix database upgrade failures while upgrading the debian package.
Fix inconsistent player list bug.
Fix agressive deletion of files in /etc/poker-network.
    https://gna.org/bugs/?13765
Fix tourney left overs https://gna.org/bugs/index.php?13769
Fix players never kicked out of the table https://gna.org/bugs/index.php?13950
Fix extra packets when rejoining a table https://gna.org/bugs/index.php?14026

Satellite tournaments

Release 1.7.4

Support for backoffice clients.

Bug fixes

Release 1.7.3

Bug fixes

Release 1.7.2

Complete REST interface.

Add de, es, nl, it, pt, da, fi, nb, sv translations.

Proxy server with authentication capabilities.

Add test coverage.

Implement authentication plugin infrastructure and example.

Release 1.7.1 

Pre-release

Release 1.7.0

Pre-release

Release 1.6.0

Improved REST support.

Release 1.5.0

Complete JSONP support + custom session names.

User names can have a size up to 50 characters.

New packet : PacketPokerBuyInLimits(Packet):
  Semantics: the buy-in boundaries for "game_id" in the range
  ["min","max"]. An optimal buy-in is suggested in "best".

Release 1.4.0

JSONP support + session control for REST API.

Bug fixes.

Release 1.3.0

Improve stateless connections to the poker server.

Release 1.2.0

Bug fixes.

Release 1.1.1

Bug fixes.

Release 1.1.0

Multiple currencies support. 

Explain poker to dummy clients mode. 

Bug fixes.

Release 1.0.37

Tournament breaks.

Bug fixes.

Internationalization of poker-web.

Release 1.0.36

Bug fixes.

Release 1.0.35

Bug fixes.

Internationalization.

Release 1.0.34

Bug fixes.

Release 1.0.33

Freerolls.

Bug fixes.

Release 1.0.32

Bug fixes.

Client server protocol change, including packets renumbering.

Broadcast messages from the server.

Release 1.0.31

Bug fixes.

Default poker web example reworked.

Integration in desktop menus.

Release 1.0.30

Bug fixes.

Release 1.0.29

Bug fixes.

Release 1.0.28

Bug fixes.

Release 1.0.27

Bug fixes.

Release 1.0.26

Bug fixes.

Affiliate information tables.

Release 1.0.25

Bug fixes.

More unit tests (protocol + children).

Packages for etch and edgy.

Release 1.0.24

Poker2D keyboard accelerators

KDE khotkeys bindings

Bug fixes.

Release 1.0.23

Major bugfixes.

Customize multiple currencies.

Release 1.0.22

Multiple currencies implementation.

Minor bug fixes.

Complete repackaging.

Release 1.0.20

Fix sound notification bugs in poker2d.

Release 1.0.19

Invalid cast major bug fix & leaks in poker2d interface.

Release 1.0.18

Sound notifications in pokerclient2d

More non-regression tests

Portability fixes

User account creation timestamp

Release 1.0.17

Fix bugous display of raise limit in pot limit

Force theme of poker2d

Database index optimization

Bug fixes

Release 1.0.16

Major bug fix in tournaments.

Bots are more aggressive in no-limit.

Packaging fixes.

Release 1.0.15

Portability fixes.

Release 1.0.14

Muck control.

New tournaments payout structures.

Fix timeout handling bug and chat history.

Transparent proxy support.

Release 1.0.13

Fix delays and interface bugs.

Improved interface for the poker client .

Release 1.0.12

Protocol has changed.

All amounts are expressed in 1/100 of units or cents.

Client negotiate with server to start the next hand as quickly
as possible.

Bug fix in protocol detection and error report.

Select a random outfit if the outfit from the server is obsolete.

User sessions archived in database.

Release 1.0.11

Major bug fixes.

Protocol changed (version 1.0.8)

PHP based web interface (account creation, editing personal settings)
using the server SOAP interface.

Release 1.0.10

Minor distributions fix.

Release 1.0.9

Minor distribution fix.

Release 1.0.8

Bug fixes related to error handling, exit conditions,
packaging and verbosity.

Release 1.0.7

The poker2d outfit is stored on the server, the skin element
is removed from the user preferences.

Standalone poker bot program.

Protocol number is no longer bound to the package number.

Major bug fixes.

Release 1.0.6

Client upgrade system that fetches a the new version from a remote
repository.

Versioning and migration system for XML configuration files.

Upgrade system for server database.

Fix i18n bugs in SOAP/XMLRPC communications.

Fix bugs related to re-joining a game after a crash.

Release 1.0.5

PacketPokerShowdown packet sent to the client with the complete
showdown stack. 

In poker2d client, showdown notification logic where players run 
specific animations based on their net income on a given hand.

Release 1.0.4

Python thread safety bug solved.

pkg-config compatibility.

Installation without packages made easy.

Tables where the player is seated are marked in poker2d.

Release 1.0.3

Major bug fix in the client library.

Release 1.0.2

Server can be made to listen on SSL or regular TCP ports.

XML-RPC and SOAP support, either with cookie based sessions
or stateless.

Authentication related bug fixes.

Server and client SSL support.

Release 1.0.1

Initial release as a standalone project after extraction from 
http://gna.org/projects/underware



pokernetwork/pokerserver.py
#
# -*- py-indent-offset: 4; coding: utf-8; mode: python -*-
#
# Copyright (C) 2006, 2007, 2008, 2009 Loic Dachary <loic@dachary.org>
# Copyright (C) 2008 Johan Euphrosine <proppy@aminche.com>
# Copyright (C) 2004, 2005, 2006 Mekensleep <licensing@mekensleep.com>
#                                24 rue vieille du temple, 75004 Paris
#
# This software's license gives you freedom; you can copy, convey,
# propagate, redistribute and/or modify this program under the terms of
# the GNU Affero General Public License (AGPL) as published by the Free
# Software Foundation (FSF), either version 3 of the License, or (at your
# option) any later version of the AGPL published by the FSF.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
# General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program in a file in the toplevel directory called
# "AGPLv3".  If not, see <http://www.gnu.org/licenses/>.
#
# Authors:
#  Loic Dachary <loic@dachary.org>
#  Henry Precheur <henry@precheur.org> (2004)
#

from pokernetwork import log as network_log
log = network_log.getChild('server')

import sys, os
sys.path.insert(0, ".")
sys.path.insert(0, "..")

import platform
from os.path import exists

try:
    from OpenSSL import SSL ; del SSL # just imported to check for SSL
    from pokernetwork.pokerservice import SSLContextFactory
    HAS_OPENSSL=True
except Exception:
    log.inform("OpenSSL not available.")
    HAS_OPENSSL=False

from twisted.application import internet, service, app
from twisted.web import server
from twisted.python import log as twisted_log

from pokernetwork.pokernetworkconfig import Config
from pokernetwork.pokerservice import PokerTree, PokerRestTree, PokerService, IPokerFactory
from pokernetwork.pokersite import PokerSite
from pokernetwork.pokermanhole import makeService as makeManholeService

from reflogging import TwistedHandler, SingleLineFormatter
import logging

def makeService(configuration):
    settings = Config([''])
    settings.load(configuration)
    if not settings.header:
        sys.exit(1)

    serviceCollection = service.MultiService()
    poker_service = PokerService(settings)
    poker_service.setServiceParent(serviceCollection)

    poker_factory = IPokerFactory(poker_service)

    #
    # Setup Logging
    #
    log_level = int(os.environ['LOG_LEVEL']) \
        if 'LOG_LEVEL' in os.environ \
        else settings.headerGetInt("/server/@log_level")
    logger = logging.getLogger()
    handler = TwistedHandler(twisted_log.theLogPublisher)
    handler.setFormatter(SingleLineFormatter('[%(refs)s] %(message)s'))
    logger.addHandler(handler)
    if log_level and not log_level in (10, 20, 30, 40, 50):
        raise ValueError(
            "Unsupported log level %d. Supported log levels "
            "are DEBUG(10), INFO(20), WARNING(30), ERROR(40), CRITICAL(50)." % (log_level,)
        )
    logger.setLevel(log_level if log_level else logging.WARNING)

    #
    # Poker protocol (with or without SSL)
    #
    tcp_port = settings.headerGetInt("/server/listen/@tcp")
    internet.TCPServer(tcp_port, poker_factory).setServiceParent(serviceCollection)

    tcp_ssl_port = settings.headerGetInt("/server/listen/@tcp_ssl")
    if HAS_OPENSSL and tcp_ssl_port:
        internet.SSLServer(tcp_ssl_port, poker_factory, SSLContextFactory(settings)).setServiceParent(serviceCollection)

    rest_site = PokerSite(settings, PokerRestTree(poker_service))

    #
    # HTTP (with or without SLL) that implements REST
    #
    rest_port = settings.headerGetInt("/server/listen/@rest")
    if rest_port:
        internet.TCPServer(rest_port, rest_site).setServiceParent(serviceCollection)

    rest_ssl_port = settings.headerGetInt("/server/listen/@rest_ssl")
    if HAS_OPENSSL and rest_ssl_port:
        internet.SSLServer(rest_ssl_port, rest_site, SSLContextFactory(settings)).setServiceParent(serviceCollection)

    http_site = server.Site(PokerTree(poker_service))

    #
    # HTTP (with or without SLL) that implements XML-RPC and SOAP
    #
    http_port = settings.headerGetInt("/server/listen/@http")
    if http_port:
        internet.TCPServer(http_port, http_site).setServiceParent(serviceCollection)

    http_ssl_port = settings.headerGetInt("/server/listen/@http_ssl")
    if HAS_OPENSSL and http_ssl_port:
        internet.SSLServer(http_ssl_port, http_site, SSLContextFactory(settings)).setServiceParent(serviceCollection)

    #
    # SSh twisted.conch.manhole
    #
    manhole_port = settings.headerGetInt("/server/listen/@manhole")
    if manhole_port:
        manhole_service = makeManholeService(manhole_port, {
            'poker_service': poker_service,
            'poker_site': rest_site
        })
        manhole_service.name = 'manhole'
        manhole_service.setServiceParent(serviceCollection)
        log.warn(
            "PokerManhole: manhole is useful for debugging, however, "
            "it can be a security risk and should be used only during debugging"
        )

    return serviceCollection

def makeApplication(argv):
    default_path = "/etc/poker-network" + sys.version[:3] + "/poker.server.xml"
    if not exists(default_path):
        default_path = "/etc/poker-network/poker.server.xml"
    configuration = argv[-1][-4:] == ".xml" and argv[-1] or default_path    
    application = service.Application('poker')
    serviceCollection = service.IServiceCollection(application)
    poker_service = makeService(configuration)
    poker_service.setServiceParent(serviceCollection)
    return application

def run():
    twisted_log.startLogging(sys.stdout)
        
    if platform.system() != "Windows":
        if 'twisted.internet.reactor' not in sys.modules:
            log.debug("installing epoll reactor")
            from twisted.internet import epollreactor
            epollreactor.install()
        else:
            log.debug("reactor already installed")
    from twisted.internet import reactor
    application = makeApplication(sys.argv)
    app.startApplication(application, None)
    reactor.run()

if __name__ == '__main__':
    # Does not need coverage since we call run directly in the tests.
    run() # pragma: no cover
    
